### PR TITLE
Fix a bug where malformed paths don't get written to the destination dir

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/kubernetes/kubernetes/issues/61297

@liggitt @hansmi @cjcullen @jessfraz

### CODE REVIEW

**Comment by cjcullen** [comment]
I would have assumed that a `srcdir/../../somefile` in the tar would cause an error instead of silently getting flattened into `destdir/somefile`. But if this is canonical, I think I'm fine with it.

**Comment by brendandburns** [comment]
@cjcullen I'm not sure what is canonical or not, this seems to me to be fine, but I can also return an error if that is the preferred approach.

File: pkg/kubectl/cmd/cp.go [hunk 1]
```diff
@@ -312,6 +312,12 @@ func recursiveTar(srcBase, srcFile, destBase, destFile string, tw *tar.Writer) e
 	return nil
 }
 
+// clean prevents path traversals by stripping them out.
+// This is adapted from https://golang.org/src/net/http/fs.go#L74
+func clean(fileName string) string {
+	return path.Clean(string(os.PathSeparator) + fileName)
```
Comment (dims) [inline pkg/kubectl/cmd/cp.go]: @brendanburns will prefixing `os.PathSeparator` work for you on windows? (is fileName likely to be `c:/abc`)?

Comment (liggitt) [reply pkg/kubectl/cmd/cp.go]: Not with any legitimate response from the container (since we're invoking the tar command).

Comment (dims) [reply pkg/kubectl/cmd/cp.go]: Ack @liggitt thanks!

**Comment by dims** [comment]
LGTM 👍

**Pull request merged**